spec.bs

```diff
@@ -289,11 +335,22 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 
 A {{URLPattern}} is associated with multiple <dfn>component</dfn> [=structs=].
 
-A [=component=] has an associated <dfn for=component>pattern string</dfn>, a [=pattern string/well formed=] [=/pattern string=], initially null.
+A [=component=] has an associated <dfn for=component>pattern string</dfn>, a [=pattern string/well formed=] [=/pattern string=], which must be set upon creation.
 
-A [=component=] has an associated <dfn for=component>regular expression</dfn>, a {{RegExp}}, initially null.
+A [=component=] has an associated <dfn for=component>regular expression</dfn>, a {{RegExp}}, which must be set upon creation.
 
-A [=component=] has an associated <dfn for=component>group name list</dfn>, an array of strings, initially an empty array.
+A [=component=] has an associated <dfn for=component>group name list</dfn>, an array of strings, which must be set upon creation.
+
+<div algorithm>
+  To <dfn>compile a component</dfn> given a string |input|, [=/options=] |options|, and [=/encoding callback=] |encoding callback|:
+
+  1. Let |part list| be the result of running [=parse a pattern string=] given |input|, |options|, and |encoding callback|.
+  1. Let (|regular expression string|, |name list|) be the result of running [=generate a regular expression and name list=] given |part list| and |options|.
+  1. Let |regular expression| be [$RegExpCreate$](|regular expression string|, "`u`").
+  1. If |regular expression| is an [=abrupt completion=], throw a {{TypeError}}.
```

Comment:
I suggest instead adding a line to the previous bullet which says "If this throws an exception, catch it, and throw a TypeError".

(Or, it might be nicer to just let it throw the original SyntaxError? That seems a bit more correct anyway.)

Reasoning: when dealing with ES abstract operations, you can either do what the ES spec does, and treat them as always returning completion values, or you can treat them like web platform algorithms, which throw exceptions/return values. If you do the ES spec way (which is done by Streams, Web IDL, and some sections of HTML) then you need to do some more hoops, e.g. to extract the actual return value you'd need to get `|regular expression|.[[Value]]`, and you'd need to update other parts of your spec to use the ! or ? prefixes.

Reply:
Allowing the original error out will probably give an error with a hard to understand message string.  The caller won't be able to see the compiled regexp string, so I don't know that it is useful.

Also, the implementation today throws a TypeError like we do for other invalid patterns.  Are you suggesting those should all be syntax errors?

Reply:
Yeah, fair point about the original message.

SyntaxError vs. TypeError is an interesting (although not super-important) question. TypeError seems good for now, but I've asked in the TC39 matrix channel what they think. It seems like sometimes the ES spec does use SyntaxError for generic "parsing errors", e.g. in JSON.parse or when you do `BigInt("NaN")`, as well as in RegExp parsing.

Reply:
Regardless of which way you go here, you should fix the editorial issue here with abrupt completion language vs. exception language.

TC39 people said this seems like a good use case for `{{SyntaxError}}` instead of `{{TypeError}}`, so that might be a good thing to change, either here or in a followup coordinated with the implementation/tests.

Reply:
Tried to fix.  Filed #74 to follow-up on exception type.